src/arrowc/arrowc.h

```diff
@@ -201,6 +275,147 @@ ArrowErrorCode ArrowSchemaSetName(struct ArrowSchema* schema, const char* name);
 /// ArrowSchemaDeepCopy.
 ArrowErrorCode ArrowSchemaSetMetadata(struct ArrowSchema* schema, const char* metadata);
 
+/// \brief Allocate the schema->children array
+///
+/// Includes the memory for each child struct ArrowSchema.
+/// schema must have been allocated using ArrowSchemaInit or
+/// ArrowSchemaDeepCopy.
+ArrowErrorCode ArrowSchemaAllocateChildren(struct ArrowSchema* schema,
+                                           int64_t n_children);
+
+/// \brief Allocate the schema->dictionary member
+///
+/// schema must have been allocated using ArrowSchemaInit or
+/// ArrowSchemaDeepCopy.
+ArrowErrorCode ArrowSchemaAllocateDictionary(struct ArrowSchema* schema);
+
+/// \brief Iterate over key/value pairs in schema metadata
+ArrowErrorCode ArrowMetadataWalk(const char* metadata,
+                                 ArrowErrorCode (*callback)(struct ArrowStringView* key,
+                                                            struct ArrowStringView* value,
+                                                            void* private_data),
+                                 void* private_data);
```

Comment:
I wonder if instead of a callback, this could be a function that you call repeatedly with an out parameter (or two) it uses for tracking where it is in the string? Callbacks might be a little annoying to work with